chat preset: GLM-4

kallewoof opened the pull request:
Chat preset for [GLM-4](https://huggingface.co/THUDM/GLM-4-32B-0414). The `[gMASK]<sop>` part in the story string is necessary for completion endpoint.

The chat template (cleaned up) looks like this:
```jinja
[gMASK]<sop>
{%- for msg in messages %}
  {%- if msg.role == 'system' %}
    <|system|>
    {{ msg.content }}
  {%- endif %}
{%- endfor %}
{%- for message in messages if message.role != 'system' %}
  {%- set role = message['role'] %}
  {%- set content = message['content'] %}
  {%- set meta = message.get(\"metadata\", \"\") %}
  {%- if role == 'user' %}
    <|user|>
    {{ content }}
  {%- elif role == 'assistant' and not meta %}
    <|assistant|>
    {{ content }}
  {%- elif role == 'assistant' and meta %}
    <|assistant|>{{ meta }} 
    {{ content }}
  {%- elif role == 'observation' %}
    <|observation|>
    {{ content }}
  {%- endif %}
{%- endfor %}
{% if add_generation_prompt %}
  <|assistant|>
{% endif %}
```

## Checklist:

- [X] I have read the [Contributing guidelines](https://github.com/SillyTavern/SillyTavern/blob/release/CONTRIBUTING.md).

kallewoof committed:
chat preset: GLM-4

Cohee1207 commented:
`[gMASK]<sop` is a BOS token? If yes, it should ideally be added by the backend.

kallewoof commented:
Yeah, hm. This is in the initial response when connecting and fetching properties:
```
srv  log_server_r: response: {"default_generation_settings":{"id":0,"id_task":-1,"n_ctx":32768,"speculative":false,"is_processing":false,"non_causal":false,"params":{"n_predict":-1,"seed":4294967295,"temperature":0.800000011920929,"dynatemp_range":0.0,"dynatemp_exponent":1.0,"top_k":40,"top_p":0.949999988079071,"min_p":0.05000000074505806,"xtc_probability":0.0,"xtc_threshold":0.10000000149011612,"typical_p":1.0,"repeat_last_n":64,"repeat_penalty":1.0,"presence_penalty":0.0,"frequency_penalty":0.0,"dry_multiplier":0.0,"dry_base":1.75,"dry_allowed_length":2,"dry_penalty_last_n":32768,"dry_sequence_breakers":["\n",":","\"","*"],"mirostat":0,"mirostat_tau":5.0,"mirostat_eta":0.10000000149011612,"stop":[],"max_tokens":-1,"n_keep":0,"n_discard":0,"ignore_eos":false,"stream":true,"logit_bias":[],"n_probs":0,"min_keep":0,"grammar":"","grammar_lazy":false,"grammar_triggers":[],"preserved_tokens":[],"chat_format":"Content-only","samplers":["penalties","dry","top_k","typ_p","top_p","min_p","xtc","temperature"],"speculative.n_max":16,"speculative.n_min":0,"speculative.p_min":0.75,"timings_per_token":false,"post_sampling_probs":false,"lora":[]},"prompt":"","next_token":{"has_next_token":true,"has_new_line":false,"n_remain":-1,"n_decoded":0,"stopping_word":""}},"total_slots":1,"model_path":"/llm/gguf/THUDM_GLM-4-32B-0414-Q8_0.gguf","chat_template":"[gMASK]<sop>{%- if tools -%}<|system|>\n# 可用工具\n{% for tool in tools %}{%- set function = tool.function if tool.get(\"function\") else tool %}\n\n## {{ function.name }}\n\n{{ function | tojson(indent=4, ensure_ascii=False) }}\n在调用上述函数时，请使用 Json 格式表示调用的参数。{%- endfor %}{%- endif -%}{%- for msg in messages %}{%- if msg.role == 'system' %}<|system|>\n{{ msg.content }}{%- endif %}{%- endfor %}{%- for message in messages if message.role != 'system' %}{%- set role = message['role'] %}{%- set content = message['content'] %}{%- set meta = message.get(\"metadata\", \"\") %}{%- if role == 'user' %}<|user|>\n{{ content }}{%- elif role == 'assistant' and not meta %}<|assistant|>\n{{ content }}{%- elif role == 'assistant' and meta %}<|assistant|>{{ meta }} \n{{ content }}{%- elif role == 'observation' %}<|observation|>\n{{ content }}{%- endif %}{%- endfor %}{% if add_generation_prompt %}<|assistant|>{% endif %}","bos_token":"[gMASK]","eos_token":"<|user|>","build_info":"b5180-13b45488"}
```

I'm not sure how to check the exact prompt used by llama-server, but the verbose output shows this, with the current config:
```
slot update_slots: id  0 | task 0 | new prompt, n_ctx_slot = 32768, n_keep = 0, n_prompt_tokens = 5982
slot update_slots: id  0 | task 0 | prompt token   0: 151331 '[gMASK]'
slot update_slots: id  0 | task 0 | prompt token   1: 151333 '<sop>'
slot update_slots: id  0 | task 0 | prompt token   2:    198 '
'
slot update_slots: id  0 | task 0 | prompt token   3: 151335 '<|system|>'
```
The above looks like it is not adding any double BOS tokens or anything. The final `log_server_r: request:` log also looks fine:
```
srv  log_server_r: request:  {"prompt":"[gMASK]<sop>\n<|system|>...
```

I do think I may need to tweak the newlines to match the tokenizer_config though, btw. I'm honestly not sure about the BOS stuff.

Edit: Looking at the template, for the case with no system message, you would still want the `[gMASK]<sop>` stuff, so `system_sequence_prefix` is probably not the right place for this, ultimately.

Also, I think removing the [gMASK] part is the right thing, as you suggest. Updating but keeping < sop > for now.

kallewoof commented:
I put < sop >\n in the story string, which seems like it does what you'd expect. Removed the [gMASK] part. Also added \n's after the user and assistant tokens.

Cohee1207 commented:
Ok thanks. if you can confirm that the prompts coming out are close to what the jinja is formatting I can merge right away.

kallewoof commented:
Sorry for delay. Will check this today and report back.

kallewoof commented:
OK, I am officially confused.

I compared to Gemma 3. The llama side has 
```
llama_model_loader: - kv  35:               tokenizer.ggml.add_bos_token bool             = true
```
and Silly sends a "< start_of_turn >etcetc" which has < bos > prepended to it in the final prompt in llama-server:
```
slot update_slots: id  0 | task 0 | prompt token   0:      2 '<bos>'
slot update_slots: id  0 | task 0 | prompt token   1:    105 '<start_of_turn>'
slot update_slots: id  0 | task 0 | prompt token   2:   2364 'user'
slot update_slots: id  0 | task 0 | prompt token   3:    107 '
'
slot update_slots: id  0 | task 0 | prompt token   4:   3048 'You'
slot update_slots: id  0 | task 0 | prompt token   5: 236789 '''
slot update_slots: id  0 | task 0 | prompt token   6:    500 're'
slot update_slots: id  0 | task 0 | prompt token   7:  82032 ' Coding'
slot update_slots: id  0 | task 0 | prompt token   8:  62710 ' Sense'
```

However, the GLM-4 GGUF metadata does *not* have the add_bos_token bool at all, which I assume means it's set to false. I also do not see the `[gMASK]` token in the prompt token list, as above:
```
slot update_slots: id  0 | task 0 | prompt token   0: 151331 '[gMASK]'
slot update_slots: id  0 | task 0 | prompt token   1: 151333 '<sop>'
slot update_slots: id  0 | task 0 | prompt token   2:    198 '
'
slot update_slots: id  0 | task 0 | prompt token   3: 151335 '<|system|>'
```

I wonder if llama.cpp metadata needs to have add_bos_token true.

kallewoof commented:
Digging further, I found [this](https://github.com/MetaGLM/glm-cookbook/blob/main/basic/glm_token_count.ipynb) notebook which shows an example of the token output. This agrees with the transformers AutoTokenizer implementation:
```python
>>> tokenizer.apply_chat_template([{"role":"user","content":"hi"},{"role":"assistant","content":"ye?"}])
[151331, 151333, 151336, 198, 6023, 151337, 198, 9008, 30]
>>> tokenizer.apply_chat_template([{"role":"user","content":"hi"},{"role":"assistant","content":"ye?"}], tokenize=False)
'[gMASK]<sop><|user|>\nhi<|assistant|>\nye?'
```

Either way, this is a llama.cpp issue I think, so I'll open a PR there. Feel free to merge this, or wait for my other PR to resolve.

Oh wait, [someone beat me to it](https://github.com/ggml-org/llama.cpp/pull/13099).

kallewoof commented:
Update: llama.cpp does not add the BOS token. Instead it is included in the chat template, so when you call the "chat completion" API endpoint it is added there. This obviously means calls to the completion endpoint needs to add this manually, so the correct thing to do seems to be to use `[gMASK]<sop>\n<|system|>...` like I originally did. Will confirm.

kallewoof commented:
Confirmed in https://github.com/ggml-org/llama.cpp/pull/13099#issuecomment-2829458947 -- updated this to do `[gMASK]<sop>` in the `story_string`.

Running against GGUF compiled with https://github.com/ggml-org/llama.cpp/pull/13099 I get:
```
slot update_slots: id  0 | task 155 | new prompt, n_ctx_slot = 32768, n_keep = 0, n_prompt_tokens = 119
slot update_slots: id  0 | task 155 | prompt token   0: 151331 '[gMASK]'
slot update_slots: id  0 | task 155 | prompt token   1: 151333 '<sop>'
slot update_slots: id  0 | task 155 | prompt token   2:    198 '
'
slot update_slots: id  0 | task 155 | prompt token   3: 151335 '<|system|>'
slot update_slots: id  0 | task 155 | prompt token   4:    198 '
'
slot update_slots: id  0 | task 155 | prompt token   5:   2610 'You'
slot update_slots: id  0 | task 155 | prompt token   6:   2299 ''re'
slot update_slots: id  0 | task 155 | prompt token   7:  47773 ' Coding'
```
which matches up with the top (fixed) output in https://github.com/ggml-org/llama.cpp/pull/13099#issuecomment-2829426830.

kallewoof commented:
I should note that running this against existing GLM-4 GGUF files (such as Bartowski's) works fine since the above fix only applies to chat completion usage. This is the output from https://huggingface.co/bartowski/THUDM_GLM-4-32B-0414-GGUF/blob/main/THUDM_GLM-4-32B-0414-Q8_0.gguf:
```
slot update_slots: id  0 | task 0 | new prompt, n_ctx_slot = 32768, n_keep = 0, n_prompt_tokens = 119
slot update_slots: id  0 | task 0 | prompt token   0: 151331 '[gMASK]'
slot update_slots: id  0 | task 0 | prompt token   1: 151333 '<sop>'
slot update_slots: id  0 | task 0 | prompt token   2:    198 '
'
slot update_slots: id  0 | task 0 | prompt token   3: 151335 '<|system|>'
slot update_slots: id  0 | task 0 | prompt token   4:    198 '
'
slot update_slots: id  0 | task 0 | prompt token   5:   2610 'You'
slot update_slots: id  0 | task 0 | prompt token   6:   2299 ''re'
slot update_slots: id  0 | task 0 | prompt token   7:  47773 ' Coding'
slot update_slots: id  0 | task 0 | prompt token   8:  45950 ' Sense'
slot update_slots: id  0 | task 0 | prompt token   9:     72 'i'
```

Cohee1207 reviewed (approved):
Ok thanks for very thorough research, I think we're done here.
